# Ex-9: Recommender Systems

## Question 1  (Recommendations by Critics) [20 points]

### Based on the example in Section 2.2 of Similarity and Recommender Systems:
### 1. [6 points] For a user named User3 who gave the following ratings:
            
            Australia: 6
            Hancock: 7
            Milk: 9
### What would the (Euclidean) distance be to each of the 6 critics?

In [ ]:
def EuclideanDist(user_rating, reviewer_rating):
    """
    Calculate the Euclidean distance between two rating vectors.
    :param user_rating: A list of ratings given by a user.
    :param reviewer_rating: A list of ratings given by a reviewer.
    :return: The Euclidean distance between the two rating vectors.
    """
    return sum((user_rating[i] - reviewer_rating[i]) ** 2 for i in range(len(user_rating))) ** 0.5

##User and Reviewer Ratings
user_rating = [6, 7, 9]
denby_scores = [3, 9, 9]
mccarthy_ratings = [7, 3, 8]
morgenstern_ratings = [7, 0, 8]
puig_ratings = [5, 5, 9]
travers_ratings = [5, 8, 10]
turan_ratings = [7, 4, 7]

# Calculate the Euclidean distance between the user rating and each reviewer's ratings
denby_distance = EuclideanDist(user_rating, denby_scores)
mccarthy_distance = EuclideanDist(user_rating, mccarthy_ratings)
morgenstern_distance = EuclideanDist(user_rating, morgenstern_ratings)
puig_distance = EuclideanDist(user_rating, puig_ratings)
travers_distance = EuclideanDist(user_rating, travers_ratings)
turan_distance = EuclideanDist(user_rating, turan_ratings)






In [6]:
print("Denby distance: ", denby_distance)
print("McCarthy distance: ", mccarthy_distance)
print("Morgenstern distance: ", morgenstern_distance)
print("Puig distance: ", puig_distance)
print("Travers distance: ", travers_distance)
print("Turan distance: ", turan_distance)

Denby distance:  3.605551275463989
McCarthy distance:  4.242640687119285
Morgenstern distance:  7.14142842854285
Puig distance:  2.23606797749979
Travers distance:  1.7320508075688772
Turan distance:  3.7416573867739413


### 2. [6 points] What is the similarity of User3 to each critic?

In [8]:
def similarity(user_rating, reviewer_rating):
    """
    Calculate the similarity between two rating vectors.
    :param user_rating: A list of ratings given by a user.
    :param reviewer_rating: A list of ratings given by a reviewer.
    :return: The similarity between the two rating vectors.
    """
    return 1 / (1 + EuclideanDist(user_rating, reviewer_rating))

In [10]:
similarity_denby = similarity(user_rating, denby_scores)
similarity_mccarthy = similarity(user_rating, mccarthy_ratings)
similarity_morgenstern = similarity(user_rating, morgenstern_ratings)
similarity_puig = similarity(user_rating, puig_ratings)
similarity_travers = similarity(user_rating, travers_ratings)
similarity_turan = similarity(user_rating, turan_ratings)
print("Denby similarity: ", similarity_denby)
print("McCarthy similarity: ", similarity_mccarthy)
print("Morgenstern similarity: ", similarity_morgenstern)   
print("Puig similarity: ", similarity_puig)
print("Travers similarity: ", similarity_travers)
print("Turan similarity: ", similarity_turan)

Denby similarity:  0.21712927295533244
McCarthy similarity:  0.1907435698305462
Morgenstern similarity:  0.122828568570857
Puig similarity:  0.3090169943749474
Travers similarity:  0.36602540378443865
Turan similarity:  0.21089672205953397


### 3. [4 points] Which critic is most similar to User3?

Morgenstern is the critic that is most similar to User3, with a similarity of 0.123

### 4. [4 points] Which film should be recommended to User3?

Using method 1 as outlined in the textbook, the recommendation would be Burn After Reading or Body of Lies, because they are tied for the highest rated film by Morenstern that User3 has not already reviewed. 